### PR TITLE
Date/Time parsing: Use java time API instead of exception handling

### PR DESCRIPTION
**Pull request by spinscale**
when several formatters are used, the existing way of parsing those is
to throw an exception catch it, and try the next one. This is is
considerably slower than the approach taken in joda time, so that
indexing throughput is reduced when a date format like `x||y` is used and y is the
date format being used.

This commit now uses the java API to parse the date by appending the
date time formatters to each other and does not rely on exception
handling.

This also removes the `MergedDateFormatter` class, is this kind of 
work can now be done before the class is created.

The stats of the added benchmark before this patch

```
Benchmark                             Mode  Cnt     Score     Error  Units
DateFormatterBenchmark.parseJavaDate  avgt   30  5638,653 ± 135,143  ns/op
DateFormatterBenchmark.parseJodaDate  avgt   30  1094,825 ±   8,526  ns/op
```

with this patch

```
Benchmark                             Mode  Cnt     Score   Error  Units
DateFormatterBenchmark.parseJavaDate  avgt   30   813,562 ± 6,450  ns/op
DateFormatterBenchmark.parseJodaDate  avgt   30  1091,707 ± 7,008  ns/op
```

I also played around a bit with the benchmark using `parseJoda` for joda, but there are no real runtime changes.

### CODE REVIEW

**Comment by spinscale** [comment]
@elasticmachine please retest this

**Comment by spinscale** [comment]
@elasticmachine please retest this

**Comment by spinscale** [comment]
this requires more investigation, parsing seems to be broken, when using more than one parser, only the first one is applied correctly

**Comment by spinscale** [comment]
@elasticmachine retest this please

File: server/src/main/java/org/elasticsearch/common/time/DateFormatters.java [hunk 1]
```diff
@@ -1483,90 +1449,22 @@ public static DateFormatter forPattern(String input) {
...
-                                + pattern + "]", e);
-                        } else {
-                            failure.addSuppressed(e);
-                        }
-                    }
-                }
-                throw failure;
-            };
-        }
+        return new JavaDateFormatter(pattern, printer, dateTimeFormatters.toArray(new DateTimeFormatter[]{}));
```
Comment (danielmitterdorfer) [inline server/src/main/java/org/elasticsearch/common/time/DateFormatters.java]: While we use both styles in our code base, in the majority of cases we'd use `dateTimeFormatters.toArray(new DateTimeFormatter[0])`.

**Review by danielmitterdorfer — COMMENTED** [review]
Looks fine overall, I left a few minor comments.

File: benchmarks/src/main/java/org/elasticsearch/benchmark/time/DateFormatterBenchmark.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+@State(Scope.Benchmark)
+@SuppressWarnings("unused") //invoked by benchmarking framework
+public class DateFormatterBenchmark {
+
+    private final DateFormatter javaFormatter = DateFormatter.forPattern("8year_month_day||ordinal_date||epoch_millis");
+    private final DateFormatter jodaFormatter = Joda.forPattern("year_month_day||ordinal_date||epoch_millis");
+
+    @Benchmark
+    public void parseJavaDate() {
+        javaFormatter.parse("1234567890");
```
[the remark this answers is not in the document]
Comment (spinscale) [reply benchmarks/src/main/java/org/elasticsearch/benchmark/time/DateFormatterBenchmark.java]: oooh, good one, thanks! fixed

File: server/src/main/java/org/elasticsearch/common/time/DateFormatters.java (hunk 1 above)
Comment (spinscale) [reply server/src/main/java/org/elasticsearch/common/time/DateFormatters.java]: done

File: server/src/main/java/org/elasticsearch/common/time/DateFormatters.java [hunk 3]
```diff
@@ -1483,90 +1449,22 @@ public static DateFormatter forPattern(String input) {
...
-                        failure = new IllegalArgumentException("failed to parse date field [" + input + "] with format [" + pattern + "]",
-                            e);
-                    } else {
-                        failure.addSuppressed(e);
-                    }
-                }
+    static JavaDateFormatter merge(String pattern, List<DateFormatter> formatters) {
+        assert formatters.size() > 0;
+
+        List<DateTimeFormatter> dateTimeFormatters = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (spinscale) [reply server/src/main/java/org/elasticsearch/common/time/DateFormatters.java]: ++ done

**Comment by spinscale** [comment]
thanks! applied your comments

**Comment by spinscale** [comment]
@elasticmachine retest this please

**Review by danielmitterdorfer — APPROVED** [review]
Thanks for iterating. Can you please rerun the microbenchmarks with the new changes and update the description? LGTM if Jenkins is also happy.

**Comment by spinscale** [comment]
there was no significant change in the benchmark stats, thus I didnt update

**Review by rjernst — APPROVED** [review]
LGTM

**Pull request merged**